### PR TITLE
PowerPC, System Z: do not reset backtrace_pos in caml_raise_exception

### PR DESCRIPTION
**Pull request by xavierleroy**
On PowerPC and System Z, caml_raise_exception would set backtrace_pos
to 0 before calling caml_stash_backtrace.

This caused the backtrace to be truncated in situations like the following:
C code calls back into OCaml code, which raises an exception
(-> backtrace is saved thanks to caml_raise_exn calling caml_stash_backtrace)
then caml_callback raises the exception again via caml_raise_exception.

The test lib-dynlink-initializers/test10_main.ml exhibited the problem.
The backtrace was wrong, and is now correct after this patch.

Other ports do not reset backtrace_pos in caml_raise_exception.
This may have been code from an early implementation of native-code
backtraces that was left over (by mistake) in the PowerPC port,
which served as a basis for the System Z port.

### CODE REVIEW

**Comment by nojb** [comment]
I guess this fixes https://github.com/ocaml/ocaml/pull/9183#issuecomment-591328185.

Thanks! I had investigated the issue a bit but was stuck.

**Comment by xavierleroy** [comment]
@nojb: yes, it does fix the issue exposed by #9183.  Could you please review and accept this PR?  Maybe then I'll forgive you for having broken Inria's CI for two months.

**Review by nojb — APPROVED** [review]
I have read the patch: it does what is claimed and indeed aligns the code for these two architectures with the rest.

Just in case, I am running a round of `precheck`: https://ci.inria.fr/ocaml/job/precheck/349/ and will merge once it passes.

**Comment by Octachron** [comment]
If you want some result from the past future, there is also https://ci.inria.fr/ocaml/job/precheck/348/ .

**Comment by nojb** [comment]
> If you want some result from the past future, there is also https://ci.inria.fr/ocaml/job/precheck/348/ .

Thanks, I don't know who I missed it...

**Comment by nojb** [comment]
@xavierleroy As this is a bug fix, could you add a Changes entry? Thanks!

**Comment by xavierleroy** [comment]
I added a Changes entry.  Please squash before merging.

**Comment by nojb** [comment]
Merged, thanks!

**Pull request merged**
